sdk/eventgrid/azure-messaging-eventgrid/src/main/java/com/azure/messaging/eventgrid/EventGridPublisherAsyncClient.java

```diff
@@ -0,0 +1,164 @@
+    private final String hostname;
+
+    private final EventGridPublisherClientImpl impl;
+
+    private final EventGridServiceVersion serviceVersion;
+
+    EventGridPublisherAsyncClient(HttpPipeline pipeline, String hostname, SerializerAdapter serializerAdapter,
+                                  EventGridServiceVersion serviceVersion) {
+        this.impl = new EventGridPublisherClientImplBuilder()
+            .pipeline(pipeline)
+            .serializerAdapter(serializerAdapter)
+            .buildClient();
+
+        // currently the service version is hardcoded into the Impl client, but once another service version gets
+        // released we should add this to the impl builder options
+        this.serviceVersion = serviceVersion;
+
+        this.hostname = hostname;
+    }
+
+    /**
+     * Get the service version of the Rest API.
+     * @return the Service version of the rest API
+     */
+    public EventGridServiceVersion getServiceVersion() {
+        return this.serviceVersion;
+    }
+
+    /**
+     * Publishes the given EventGrid events to the set topic or domain.
+     * @param events the EventGrid events to publish.
+     *
+     * @return the completion.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Void> sendEvents(Iterable<EventGridEvent> events) {
+        return withContext(context -> sendEvents(events, context));
+    }
+
+    Mono<Void> sendEvents(Iterable<EventGridEvent> events, Context context) {
+        return Flux.fromIterable(events)
+            .map(EventGridEvent::toImpl)
+            .collectList()
+            .flatMap(list -> this.impl.publishEventsAsync(this.hostname, list, context));
+    }
+
+    /**
+     * Publishes the given cloud events to the set topic or domain.
+     * @param events the cloud events to publish.
+     *
+     * @return the completion.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Void> sendCloudEvents(Iterable<CloudEvent> events) {
+        return withContext(context -> sendCloudEvents(events, context));
+    }
+
+    Mono<Void> sendCloudEvents(Iterable<CloudEvent> events, Context context) {
+        return Flux.fromIterable(events)
+            .map(CloudEvent::toImpl)
+            .collectList()
+            .flatMap(list -> this.impl.publishCloudEventEventsAsync(this.hostname, list, context));
+    }
+
+    /**
+     * Publishes the given custom events to the set topic or domain.
+     * @param events the custom events to publish.
+     *
+     * @return the completion.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Void> sendCustomEvents(Iterable<Object> events) {
+        return withContext(context -> sendCustomEvents(events, context));
+    }
+
+    Mono<Void> sendCustomEvents(Iterable<Object> events, Context context) {
+        return Flux.fromIterable(events)
+            .collectList()
+            .flatMap(list -> this.impl.publishCustomEventEventsAsync(this.hostname, list, context));
```

Comment:
Unfortunately, we'll have to iterate through the collection twice - once to convert Iterable to list and then when serializing. This impacts performance. Not sure, if we can avoid this without changing the API to take a `List` instead.